Enhance Hero widget with customizable animation curves

aaazlkm opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

Follow-up to #159393 (closed due to inactivity).
Implements the same feature with changes based on the previous review feedback.

Related issue: https://github.com/flutter/flutter/issues/26150

## Screenshot

Using elasticOut because the difference was hard to see with fastOutSlowIn.
<table>
  <tr>
    <th>curve=elasticOut, reverseCurve=null</th>
    <th>curve=elasticOut, reverseCurve=linear</th>
  </tr>
  <tr>
    <td>
      <video src="https://github.com/user-attachments/assets/c8c39fb0-e677-471e-ac2a-2a062277ae48"
             controls muted loop width="320"></video>
    </td>
    <td>
      <video src="https://github.com/user-attachments/assets/1f39d0a8-ed80-4404-b766-38844cca3909"
             controls muted loop width="320"></video>
    </td>
  </tr>
</table>

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

aaazlkm committed:
Enhance Hero widget with customizable animation curves

- Added `curve` and `reverseCurve` properties to the Hero widget to allow customization of animation curves during transitions.
- Updated the animation logic to utilize the specified curves for both forward and reverse transitions.
- Enhanced tests to verify the behavior of customizable curves, ensuring that the animations reflect the specified curves correctly.

aaazlkm commented on packages/flutter/test/widgets/heroes_test.dart:
```diff
@@ -620,47 +620,111 @@ Future<void> main() async {
+      final observer = TransitionDurationObserver();
+      await tester.pumpWidget(
+        MaterialApp(
+          key: UniqueKey(),
```
Recreate MaterialApp with a unique key to avoid conflicts from previous curve tests.

aaazlkm replied:
Thank you for the review! I updated the code as you suggested. Both the code and the test output are now easier to read—this was really helpful!
https://github.com/flutter/flutter/pull/180100/changes/3701d0db6e9e83cdffe18bd6f48e0390a1045181

aaazlkm commented on packages/flutter/lib/src/widgets/heroes.dart:
```diff
@@ -459,10 +471,23 @@ class _HeroFlightManifest {
+        parent = fromRoute.animation!;
+        curve = fromHero.widget.curve;
+        reverseCurve = (fromHero.widget.reverseCurve ?? curve).flipped;
+    }
```
discussion on https://github.com/flutter/flutter/pull/180100#discussion_r2632969761

aaazlkm committed:
Refactor hero tests for improved readability

- Simplified the calls to `verifyPoppedHeroCurve` by removing unnecessary line breaks, enhancing code clarity.
- Adjusted the `await tester.pump` line for consistent formatting in the test cases.

aaazlkm committed:
Clarified the behavior of the `reverseCurve` property in the Hero widget's documentation to specify that if it is null,

aaazlkm committed:
Bring back the original ‘Heroes animation is fastOutSlowIn’ test.

aaazlkm committed:
Refactor reverseCurve handling in Hero animations

- Updated the logic for the `reverseCurve` property in the Hero widget to simplify the assignment, ensuring that the flipped curve is applied only when necessary.
- Adjusted related test cases to reflect the new handling of `reverseCurve`, maintaining consistency in animation behavior.

aaazlkm commented on packages/flutter/lib/src/widgets/heroes.dart:
```diff
@@ -261,6 +263,16 @@ class Hero extends StatefulWidget {
+
+  /// The curve to use in the reverse direction.
+  ///
+  /// If this property is null, the flipped version of [curve] is used.
```
Fixed in https://github.com/flutter/flutter/pull/180100/changes/fdd9db6bb91cfe5a81a11795d69d08b35c333908

aaazlkm commented on packages/flutter/test/widgets/heroes_test.dart:
```diff
@@ -620,47 +620,111 @@ Future<void> main() async {
     await tester.pump(const Duration(seconds: 1));
   });
 
-  testWidgets('Heroes animation is fastOutSlowIn', (WidgetTester tester) async {
```
Fixed in https://github.com/flutter/flutter/pull/180100/changes/6a568d5785c334ac017ce8cd008222caa3b4eb3a

aaazlkm committed:
Bring back the original ‘popped hero uses fastOutSlowIn curve’ test.

victorsanni commented on packages/flutter/test/widgets/heroes_test.dart:
```diff
@@ -3258,6 +3365,146 @@ Future<void> main() async {
     expect(heroSize, tween.transform(1.0));
   });
 
+  Future<void> verifyPoppedHeroCurve({
```
Our convention is to define helper methods at the very end/top of the file and outside the `main` method, not between `testWidgets` calls.

aaazlkm replied:
Fixed in https://github.com/flutter/flutter/pull/180100/changes/bfbbe1fb4a808c9df66fff46805a37dd6e971184

aaazlkm committed:
Update documentation for Hero widget's animation curves

- Clarified the default values for the `curve` and `reverseCurve` properties in the Hero widget documentation, specifying that they default to [Curves.fastOutSlowIn] and [Hero.curve].flipped respectively.

aaazlkm committed:
Add verifyPoppedHeroCurve function to enhance Hero widget tests

- Introduced a new test utility function, `verifyPoppedHeroCurve`, to validate the animation behavior of the Hero widget with customizable curves.
- The function tests the size transitions of the Hero widget during push and pop operations, ensuring that the specified curves are applied correctly throughout the animation.
- Removed the previous implementation of `verifyPoppedHeroCurve` to streamline the test code.

aaazlkm committed:
Merge remote-tracking branch 'upstream/master' into issue-26150

aaazlkm commented:
@victorsanni Thanks! I’ve applied your latest comments and pushed an update. Could you take another look?

justinmc reviewed (approved):
LGTM with a nit, but I'll defer to @victorsanni.

aaazlkm committed:
Refactor Hero widget tests for customizable animation curves

- Removed unnecessary UniqueKey from MaterialApp in tests.
- Simplified the test structure for customizable hero animation curves by using inline lists instead of separate variables.
- Enhanced test descriptions to clarify the specific curves being tested for both normal and reversed animations.
- Ensured consistency in testing the behavior of the Hero widget with various animation curves.

aaazlkm committed:
Refactor testWidgets for Heroes animation curves

- Reformatted testWidgets for better readability by splitting parameters across multiple lines.
- Maintained clarity in test descriptions for customizable animation curves in the Hero widget.
- Ensured consistent formatting across all related test cases.

chunhtai reviewed (approved):
LGTM
